spock-core/src/main/java/org/spockframework/mock/runtime/InteractionScope.java

```diff
@@ -20,17 +20,22 @@
 import org.spockframework.util.ExceptionUtil;
 
 import java.util.*;
+import java.util.concurrent.ArrayBlockingQueue;
+import java.util.concurrent.Semaphore;
 
 /**
  * A scope for interactions defined outside a then-block
  *
  * @author Peter Niederwieser
  */
 public class InteractionScope implements IInteractionScope {
+  private static final int MAX_PREVIOUS_INVOCATIONS = 5;
   private final List<IMockInteraction> interactions = new ArrayList<>();
   private final List<IMockInvocation> unmatchedInvocations = new ArrayList<>();
   private int currentRegistrationZone = 0;
   private int currentExecutionZone = 0;
+  private final Semaphore previousInvocationsUpdateSemaphore = new Semaphore(1);
```

Comment:
I don't think we need that, the MockController is already `synchronized` see org.spockframework.mock.runtime.MockController#handle
`unmatchedInvocations` is unguarded as well, you could also use an `ArrayDeque`

Reply:
Hm, I'll have a look again. I just saw it is called on different threads of course, if your SUT works multi threaded. I wasn't aware there is external synchronization